src/network_engine.cpp

```diff
@@ -0,0 +1,875 @@
+    pk.pack(std::string("y")); pk.pack(std::string("q"));
+    pk.pack(std::string("v")); pk.pack(my_v);
+
+    Blob b {buffer.data(), buffer.data() + buffer.size()};
+    std::shared_ptr<Request> req(new Request {tid.getTid(), n, std::move(b),
+        [=](std::shared_ptr<Request> req_status, ParsedMessage&& msg) { /* on done */
+            if (on_done) {
+                on_done(req_status, deserializeNodesValues(msg));
+            }
+        },
+        [=](std::shared_ptr<Request> req_status, bool) { /* on expired */
+            if (on_expired) {
+                on_expired(req_status, {});
+            }
+        }
+    });
+    sendRequest(req);
+    ++out_stats.get;
+    return req;
+}
+
+NetworkEngine::RequestAnswer
+NetworkEngine::deserializeNodesValues(ParsedMessage& msg) {
+    RequestAnswer req_a {msg.token, 0, std::move(msg.values), {}, {}};
+    if (msg.nodes4.size() % NODE4_INFO_BUF_LEN != 0 || msg.nodes6.size() % NODE6_INFO_BUF_LEN != 0) {
+        throw DhtProtocolException {DhtProtocolException::WRONG_NODE_INFO_BUF_LEN};
+    } else {
+        // deserialize nodes
+        for (unsigned i = 0; i < msg.nodes4.size() / NODE4_INFO_BUF_LEN; i++) {
+            uint8_t *ni = msg.nodes4.data() + i * NODE4_INFO_BUF_LEN;
+            const InfoHash& ni_id = *reinterpret_cast<InfoHash*>(ni);
+            if (ni_id == myid)
+                continue;
+            sockaddr_in sin;
+            std::fill_n((uint8_t*)&sin, sizeof(sockaddr_in), 0);
+            sin.sin_family = AF_INET;
+            memcpy(&sin.sin_addr, ni + ni_id.size(), 4);
+            memcpy(&sin.sin_port, ni + ni_id.size() + 4, 2);
+            if (auto n = onNewNode(ni_id, (sockaddr*)&sin, sizeof(sin), 0))
+                req_a.nodes.emplace_back(std::move(n));
+        }
+        for (unsigned i = 0; i < msg.nodes6.size() / NODE6_INFO_BUF_LEN; i++) {
+            uint8_t *ni = msg.nodes6.data() + i * NODE6_INFO_BUF_LEN;
+            const InfoHash& ni_id = *reinterpret_cast<InfoHash*>(ni);
+            if (ni_id == myid)
+                continue;
+            sockaddr_in6 sin6;
+            std::fill_n((uint8_t*)&sin6, sizeof(sockaddr_in6), 0);
+            sin6.sin6_family = AF_INET6;
+            memcpy(&sin6.sin6_addr, ni + HASH_LEN, 16);
+            memcpy(&sin6.sin6_port, ni + HASH_LEN + 16, 2);
+            if (auto n = onNewNode(ni_id, (sockaddr*)&sin6, sizeof(sin6), 0))
+                req_a.nodes.emplace_back(std::move(n));
+        }
+    }
+    return req_a;
+}
+
+void
+NetworkEngine::sendNodesValues(const sockaddr* sa, socklen_t salen, TransId tid, const Blob& nodes, const Blob& nodes6,
+        const std::vector<std::shared_ptr<Value>>& st, const Blob& token) {
+    msgpack::sbuffer buffer;
+    msgpack::packer<msgpack::sbuffer> pk(&buffer);
+    pk.pack_map(4);
+
+    pk.pack(std::string("r"));
+    pk.pack_map(2 + (not st.empty()?1:0) + (nodes.size()>0?1:0) + (nodes6.size()>0?1:0) + (not token.empty()?1:0));
```

Comment:
What do you mean? This creates a map of variable size depending on the number of fields to be packed.